auth/src/krb.cpp

```diff
@@ -0,0 +1,942 @@
+                   gss_release_buffer, instead clear it. */
+                std::memset( ikrbScratchBuffer, 0, SCRATCH_BUFFER_SIZE );
+
+                if (majorStatus != GSS_S_COMPLETE && majorStatus != GSS_S_CONTINUE_NEEDED) {
+                    constexpr bool is_client = true;
+                    const std::string msg = fmt::format(fmt::runtime(
+                                "Failed initializing KRB context. Major status: {}\tMinor status: {}"),
+                                majorStatus, minorStatus);
+                    krb_log_error(_comm.rError, msg.data(), majorStatus, minorStatus, is_client);
+                    majorStatus = gss_release_name(&minorStatus, &target_name);
+                    if (majorStatus != GSS_S_COMPLETE) {
+                        static const char* release_msg = "Error releasing name.";
+                        krb_log_error(_comm.rError, release_msg, majorStatus, minorStatus, is_client);
+                    }
+                    THROW(KRB_ERROR_INIT_SECURITY_CONTEXT, msg);
+                }
+
+                if (send_tok.length != 0) {
+                    if (const auto err = krb_send_token( &send_tok, fd ); !err.ok()) {
+                        // TODO: should we be checking these release calls?
+                        gss_release_buffer(&minorStatus, &send_tok);
+                        gss_release_name(&minorStatus, &target_name);
+                        THROW(err.code(), err.result());
+                    }
+                }
+
+                gss_release_buffer(&minorStatus, &send_tok);
+
+                if (majorStatus == GSS_S_CONTINUE_NEEDED) {
+                    recv_tok.value = &ikrbScratchBuffer;
+                    recv_tok.length = SCRATCH_BUFFER_SIZE;
+                    unsigned int bytes_read;
+                    if (const auto ret = krb_receive_token(fd, &recv_tok, &bytes_read); !ret.ok()) {
+                        gss_release_name( &minorStatus, &target_name );
+                        const auto err = PASSMSG("Error reading KRB token.", ret);
+                        THROW(err.code(), err.result());
+                    }
+
+                    tokenPtr = &recv_tok;
+                }
+            }
+            while (majorStatus == GSS_S_CONTINUE_NEEDED);
+
+            resp["next_operation"] = AUTH_CLIENT_AUTH_RESPONSE;
+
+            return resp;
+        } // kerberos_auth_establish_context
+
+        json kerberos_auth_client_response(rcComm_t& comm, const json& req)
+        {
+            namespace ie = irods::experimental;
+
+            if (auto [e, n] = verify(req, {"user_name","zone_name"}); !e) {
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in request", __func__, __LINE__, n));
+            }
+
+            json svr_req{req};
+            svr_req["next_operation"] = AUTH_AGENT_AUTH_RESPONSE;
+            auto resp = request(comm, svr_req);
+
+            comm.loggedIn = 1;
```

Comment:
Will `request()` throw if something fails or doesn't meet the auth requirements?

I'm guessing that's the only way L713 isn't executed? And that is the expected way of doing things here?

Reply:
`request` will throw an `irods::exception` if the API call returns with a code that is less than 0. This is the expected flow for the new plugin framework, yes.

Reply:
And that exception is caught by the framework. The framework will then send an appropriate error response to the client, correct?

Reply:
Not necessarily. There is a free function called `authenticate_client` which is used as the entry point for authenticating a client (similar to `clientLogin` in the legacy plugin flow). It invokes `call` on the operation for the resolved authentication plugin, and that operation will call into this function, which calls `request`, which may throw. The framework itself does not catch any exceptions.

Whoever calls `authenticate_client` is responsible for catching the exceptions. I placed the call to `authenticate_client` in `clientLogin` since this is what all of our C-based clients use to authenticate with the server. The logic for catching and swizzling a nice error response for the end user is held there: https://github.com/irods/irods/blob/6b319b807fe4e2aadb68016d8f9e2459e788cb91/lib/core/src/clientLogin.cpp#L49

Hopefully that did not make things more confusing :)

Reply:
Got it. That makes sense.